src/libs/sfdk/dockervirtualmachine.cpp

```diff
@@ -0,0 +1,529 @@
+const char STOP[] = "stop";
+const char COMMIT[] = "commit";
+const char CONTAINER_LIST[] = "ps";
+const char DETACHED[] = "-d";
+const char PUBLISH_PORT[] = "-p";
+const char VOLUME[] = "-v";
+const char CGROUP_MOUNT[] = "/sys/fs/cgroup:/sys/fs/cgroup:ro";
+const char TMPFS[] = "--tmpfs";
+const char TMPFS1[] = "/run:exec,mode=755";
+const char TMPFS2[] = "/run/lock";
+const char TMPFS3[] = "/tmp:exec";
+const char FILTER_PREFIX[] = "--filter=ancestor=";
+const char FORMAT_CONTAINER_LIST[] = "--format={{.Image}} {{.ID}}";
+const char COMMAND_INSPECT[] = "inspect";
+const char FORMAT_LABEL_LIST[] = "--format={{json .Config.Labels}}";
+const char IMAGE[] = "image";
+const char LIST[] = "ls";
+const char FORMAT_REPOSITORY[] = "--format={{.Repository}}";
+const char UNNAMED_IMAGE[] = "<none>";
+const char COMMAND_BUILD[] = "build";
+const char DOCKERFILE_PREFIX[] = "FROM ";
+const char LABEL_PREFIX[] = "--label=";
+const char TAG[] = "--tag";
+const quint16 GUESTSSHPORT = 22;
+const quint16 GUESTWWWPORT = 9292;
+}
+
+namespace Sfdk {
+
+void DockerVirtualMachinePrivate::fetchInfo(VirtualMachineInfo::ExtraInfos extraInfo,
+        const QObject *context, const Functor<const VirtualMachineInfo &, bool> &functor) const
+{
+    Q_Q(const DockerVirtualMachine);
+    Q_ASSERT(context);
+    Q_ASSERT(functor);
+
+    // TODO: fetch snapshot info if requested
+    Q_UNUSED(extraInfo)
+
+    QStringList arguments;
+    arguments.append(COMMAND_INSPECT);
+    arguments.append(q->name());
+    arguments.append(FORMAT_LABEL_LIST);
+
+    auto runner = std::make_unique<ProcessRunner>(dockerPath(), arguments);
+    QObject::connect(runner.get(), &ProcessRunner::done, q,
+                     [this, functor, process = runner->process()](bool ok) {
+        VirtualMachineInfo info;
+        if (ok)
+            info = virtualMachineInfoFromOutput(process->readAllStandardOutput().trimmed());
+
+        functor(info, ok);
+    });
+    SdkPrivate::commandQueue()->enqueue(std::move(runner));
+}
+
+void DockerVirtualMachinePrivate::start(const QObject *context, const Functor<bool> &functor)
+{
+    Q_ASSERT(context);
+    Q_ASSERT(functor);
+    QTC_ASSERT(cachedInfo().sshPort != 0, return);
+    QTC_ASSERT(cachedInfo().wwwPort != 0, return);
+    QTC_ASSERT(!cachedInfo().sharedSrc.isEmpty(), return);
+    QTC_ASSERT(!cachedInfo().sharedHome.isEmpty(), return);
+    QTC_ASSERT(!cachedInfo().sharedTargets.isEmpty(), return);
+    QTC_ASSERT(!cachedInfo().sharedConfig.isEmpty(), return);
+    QTC_ASSERT(!cachedInfo().sharedSsh.isEmpty(), return);
+    Q_Q(DockerVirtualMachine);
+
+    QStringList arguments;
+    arguments.append(RUN);
+    arguments.append(DETACHED);
+    arguments.append(PRIVILEGED1);
+    arguments.append(PRIVILEGED2);
+    arguments.append(VOLUME);
+    arguments.append(CGROUP_MOUNT);
+
+    auto addTmpfs = [&arguments](const QString& path) {
+        arguments.append(TMPFS);
+        arguments.append(path);
+    };
+
```

Comment:
nitpick: I would stick the definition and use together

Reply:
you mean <code>arguments.append("--tmpfs")</code> instead of <code>arguments.append(TMPFS)</code>?

Reply:
I mean I would omit the blank line between lambda definition and its calls here to help eyes spotting their purpose without actually reading the code :)

Reply:
fixed